hw/arm/hercules.c

```diff
@@ -0,0 +1,674 @@
+    sbd = SYS_BUS_DEVICE(&s->vim);
+    irq = qdev_get_gpio_in(DEVICE(&s->cpu), ARM_CPU_IRQ);
+    sysbus_connect_irq(sbd, 0, irq);
+    irq = qdev_get_gpio_in(DEVICE(&s->cpu), ARM_CPU_FIQ);
+    sysbus_connect_irq(sbd, 1, irq);
+    sysbus_mmio_map(sbd, 0, HERCULES_VIM_ECC_ADDR);
+    sysbus_mmio_map(sbd, 1, HERCULES_VIM_CONTROL_ADDR);
+    sysbus_mmio_map(sbd, 2, HERCULES_VIM_RAM_ADDR);
+
+    vim = DEVICE(&s->vim);
+
+    object_property_set_bool(OBJECT(&s->esm), true, "realized",
+                             &error_abort);
+    sbd = SYS_BUS_DEVICE(&s->esm);
+    sysbus_mmio_map(sbd, 0, HERCULES_ESM_ADDR);
+    irq = qdev_get_gpio_in(vim, HERCULES_ESM_HIGH_LEVEL_IRQ);
+    sysbus_connect_irq(sbd, 0, irq);
+    irq = qdev_get_gpio_in(vim, HERCULES_ESM_LOW_LEVEL_IRQ);
+    sysbus_connect_irq(sbd, 1, irq);
+
+    esm = DEVICE(&s->esm);
+
+    object_property_set_bool(OBJECT(&s->l2ramw), true, "realized",
+                             &error_abort);
+    sbd = SYS_BUS_DEVICE(&s->l2ramw);
+    sysbus_mmio_map(sbd, 0, HERCULES_RAM_ADDR);
+    sysbus_mmio_map(sbd, 1, HERCULES_L2RAMW_ADDR);
+    error = qdev_get_gpio_in(esm, HERCULES_L2RAMW_TYPE_B_UNCORRECTABLE_ERROR);
+    sysbus_connect_irq(sbd, 0, error);
+
+    object_property_set_bool(OBJECT(&s->system), true, "realized",
+                             &error_abort);
+    sbd = SYS_BUS_DEVICE(&s->system);
+    sysbus_mmio_map(sbd, 0, HERCULES_SYS_ADDR);
+    sysbus_mmio_map(sbd, 1, HERCULES_SYS2_ADDR);
+    sysbus_mmio_map(sbd, 2, HERCULES_PCR_ADDR);
+    sysbus_mmio_map(sbd, 3, HERCULES_PCR2_ADDR);
+    sysbus_mmio_map(sbd, 4, HERCULES_PCR3_ADDR);
+    sysbus_connect_irq(sbd, 0, qdev_get_gpio_in(vim, HERCULES_SSI_IRQ));
+    error = qdev_get_gpio_in(esm, HERCULES_PLL1_SLIP_ERROR);
+    sysbus_connect_irq(sbd, 1, error);
+    error = qdev_get_gpio_in(esm, HERCULES_PLL2_SLIP_ERROR);
+    sysbus_connect_irq(sbd, 2, error);
+
+    create_unimplemented_device("pinmux", HERCULES_PINMUX_ADDR,
+                                HERCULES_PINMUX_SIZE);
+
+    object_property_set_bool(OBJECT(&s->l2fmc), true, "realized",
+                             &error_abort);
+    sbd = SYS_BUS_DEVICE(&s->l2fmc);
+    sysbus_mmio_map(sbd, 0, HERCULES_L2FMC_ADDR);
+    sysbus_mmio_map(sbd, 1, HERCULES_EPC_ADDR);
+    error = qdev_get_gpio_in(esm, HERCULES_L2FMC_UNCORRECTABLE_ERROR);
+    sysbus_connect_irq(sbd, 0, error);
+    error = qdev_get_gpio_in(esm, HERCULES_CR5F_FATAL_BUS_ERROR);
+    sysbus_connect_irq(sbd, 1, error);
+    error = qdev_get_gpio_in(esm, HERCULES_EPC_CORRECTABLE_ERROR);
+    sysbus_connect_irq(sbd, 2, error);
+
+    otp_bank1_data = g_new0(uint8_t, HERCULES_OTP_BANK1_SIZE);
+    if (0) {
+        /*
+         * Handle loading OTP1 from a custom file here
+         */
+    } else {
+        /*
+         * FIXME: Convenience workaround, should eventually be removed
+         * once loading from custom file is implemented
+         */
+        const HerculesOTPData dummy_otp = {
```

Comment:
This feels like it will be an irritating thing to deal with regarding LE/BE

Reply:
Oh, this thing. When we were working on it, I eventually re-implemented it via `-device loader,file=<blah>,address=<blah>'` and plumbed it in glorious `durculess`, but you wanted to keep this dummy one for the sake of convenience. I think we should just drop this bit of code and rely on `-device loader`

Reply:
Oh, absolutely. I had forgotten the durcules launcher was overriding this anyways